cli/cmd/profile.go

```diff
@@ -118,11 +147,149 @@ Example:
 
 	cmd.PersistentFlags().BoolVar(&options.template, "template", options.template, "Output a service profile template")
 	cmd.PersistentFlags().StringVar(&options.openAPI, "open-api", options.openAPI, "Output a service profile based on the given OpenAPI spec file")
+	cmd.PersistentFlags().StringVar(&options.tap, "tap", options.tap, "Output a service profile based on tap data for the given target resource")
+	cmd.PersistentFlags().DurationVar(&options.tapDuration, "tap-duration", options.tapDuration, "Duration over which tap data is collected (for example: \"10s\", \"1m\", \"10m\")")
+	cmd.PersistentFlags().UintVar(&options.routeLimit, "route-limit", options.routeLimit, "Max number of routes to add to the profile")
 	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace of the service")
 
 	return cmd
 }
 
+func renderTapOutputProfile(options *profileOptions, controlPlaneNamespace string, w io.Writer) error {
+	profile := sp.ServiceProfile{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      fmt.Sprintf("%s.%s.svc.cluster.local", options.name, options.namespace),
+			Namespace: controlPlaneNamespace,
+		},
+		TypeMeta: meta_v1.TypeMeta{
+			APIVersion: "linkerd.io/v1alpha1",
+			Kind:       "ServiceProfile",
+		},
+	}
+
+	client := cliPublicAPIClient()
+	res, err := util.BuildResource(options.namespace, options.tap)
+	if err != nil {
+		return err
+	}
+
+	// there is kind of a duplication of param parsing, because we need to reformulate
+	// a request like linkerd profile --tap deploy/web to run the tap
+	// linkerd tap deploy --to deploy/web
+	requestParams := util.TapRequestParams{
+		Resource:    res.Type,
+		Namespace:   options.namespace,
+		ToResource:  options.tap,
+		ToNamespace: options.namespace,
+	}
+	log.Debugf("Running `linkerd tap %s  --namespace %s --to %s --to-namespace %s`", res.Type, options.namespace, options.tap, options.namespace)
+
+	req, err := util.BuildTapByResourceRequest(requestParams)
+	if err != nil {
+		return err
+	}
+	rsp, err := client.TapByResource(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	routes := routeSpecFromTap(rsp, options.tapDuration, int(options.routeLimit))
+
+	profile.Spec.Routes = routes
+	output, err := yaml.Marshal(profile)
+	if err != nil {
+		return fmt.Errorf("Error writing Service Profile: %s", err)
+	}
+	w.Write(output)
+	return nil
+}
+
+func routeSpecFromTap(tapClient pb.Api_TapByResourceClient, tapDuration time.Duration, routeLimit int) []*sp.RouteSpec {
+	routes := make([]*sp.RouteSpec, 0)
+	routesMap := make(map[string]*sp.RouteSpec)
+
+	tapEventChannel := make(chan *pb.TapEvent, 10)
+	timerChannel := make(chan struct{}, 1)
```

Comment:
I was just about to comment something similar to siggy's suggestion. 😄. in addition, I think you can also simplify the routeLimit check so that it wraps the `getPathDataFromTap(...)` method. So something like this:
```diff
diff --git a/cli/cmd/profile.go b/cli/cmd/profile.go
index 9090c951..26c90857 100644
--- a/cli/cmd/profile.go
+++ b/cli/cmd/profile.go
+               if len(routesMap) <= routeLimit {
                        routeSpec := getPathDataFromTap(event)
-
-                       if len(routesMap) > routeLimit {
-                               stopTap = true
-                               break
-                       }
-
                        if routeSpec != nil {
                                routesMap[routeSpec.Name] = routeSpec
                        }
```

Reply:
changed the structure of this per siggy's suggestion; I think the check/break is a little simpler now!